### DISCUSSION TITLE
API Error Clarification - 422 Fail to create protected rule, please ensure billing plan include protected branch gate.

### DISCUSSION THREAD

**Discussion by jakeprice-me** [community/community#30201] [Apps, API and Webhooks]
Has anybody come across the below API error before? I cannot find anything about it anywhere on the net.

> 422 Fail to create protected rule, please ensure billing plan include protected branch gate.

This happens when attempting to create a repository environment via Terraform (and/or Pulumi).

I am at a loss as to what it's trying to tell me. Any help would be appreciated.

**Comment by jakeprice-me** [comment]
A bit more info, if I run what I believe to be the underlying API call, I get "Not Found".

```sh
curl -H "Accept: application/vnd.github+json" -H "Authorization: token <my-token>" https://api.github.com/repos/<my-org>/<my-repo>/environments/<env-name>/deployment-branch-policies
```

```json
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/deployments/branch-policies#list-deployment-branch-policies"
}
```

If I take it back to just the `https://api.github.com/repos/<my-org>/<my-repo>/environments/` endpoint then I also get the above error message. However, the repo definetely exists because if I use the `https://api.github.com/repos/<my-org>/<my-repo>` endpoint I do get repository information returned.

**Comment by jakeprice-me** [comment]
I imagine I'm missing some permissions somewhere? However, I've tried it with my main account, which has owner permissions to the organisation, and still can't get it to work. No success either, if as an organisation owner I use a token with every single permission allowed.

**Comment by jakeprice-me** [comment]
I should clarify further that the environment does exist as well. I can see it on the GUI.

**Reply by byrneh** [reply]
@jakeprice-dev if you do a LIst environmets API call with valid token and scope permission to the target repository a 404 nor found would indicate there are no envirionment for that repository.  Tyhe same applies for other similar APIS responses

**Reply by jakeprice-me** [reply]
Thanks for coming back @byrneh I must have not had the right combination of scope perms and user perms. Will take another look.

**Reply by ghost** [reply]
Hitting this as well.  Assigning all available scopes to an access token does not resolve the issue.  Might create an issue in https://github.com/pulumi/pulumi-github shortly if still blocked

**Reply by jakeprice-me** [reply]
Glad to know it's not just me @thomas-gallop, that's the same behaviour I'm hitting - nothing even about the error message online and I'm not doing anything obviously wrong. I do have a support ticket raised with GitHub for more clarity on that error, when I hear back I'll post the reply here (if it's any help). I wanted to check that route before raising it with Pulumi, but feel free to. 

If you raise an issue do let me know, otherwise when I get a spare moment I'll probably raise it, if you haven't already.

**Reply by ioegcr** [reply]
I have the same behaviour, @jakeprice-dev were you able to solve the issue?

**Reply by ghost** [reply]
No solution yet, but some good info coming up in this TF issue,  https://github.com/integrations/terraform-provider-github/issues/1262

**Comment by byrneh** [comment]
Hi all, just in case any of you are using the GitHub free plan, protected branches are only supported on "Public" visibility repositories. 
If you want protected branches on "Private" repositories you will need to use a paid plan.
if it is not possible in the GUI on your target repository it will not be possible through the API.

**Reply by haploeco** [reply]
To expound on this, Environment Protection Rules on private repositories are apparently not available in the paid "Team" plan either, and is only available in the Enterprise plan.

I was looking in `https://github.com/organizations/<my_org>/billing/plans` and noticed this when I ran into this issue this morning as well. :disappointed: 

![pic-selected-230212-0729-45](https://user-images.githubusercontent.com/6033003/218317084-e1684b09-32e7-4b56-a845-f5f5adb0d872.png)

**Reply by hipages-guoxu** [reply]
sad, same issue.

**Comment by bas-ie** [comment]
This issue is also present on the API:
```
curl \
  -X PUT \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer ghp_xxx" \
  https://api.github.com/repos/anorgname/this.is.temporary/environments/test \
  -d '{"wait_timer":0,"reviewers":[{"type":"Team","id":39527}],"deployment_branch_policy":{"protected_branches":false,"custom_branch_policies":true}}'
{
  "message": "Fail to create protected rule, please ensure billing plan include protected branch gate.",
  "documentation_url": "https://docs.github.com/rest/reference/repos#create-or-update-an-environment"
}
```
A successful response can be achieved if you remove both `wait_timer` and `reviewers`, but then `deployment_branch_policy` is not set on the resulting resource:
```
curl \
  -X PUT \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer ghp_xxx" \
  https://api.github.com/repos/anorgname/this.is.temporary/environments/test \
  -d '{"deployment_branch_policy":{"protected_branches":false,"custom_branch_policies":true}}'                            
{
  "id": 657937424,
  "node_id": "EN_kwDOIIiAQ84nN1QQ",
  "name": "test",
  "url": "https://api.github.com/repos/anorgname/this.is.temporary/environments/test",
  "html_url": "https://github.com/anorgname/this.is.temporary/deployments/activity_log?environments_filter=test",
  "created_at": "2022-10-05T02:52:51Z",
  "updated_at": "2022-10-05T02:52:51Z",
  "protection_rules": [

  ]
}
```
The organisation in question is on a paid plan, and can ordinarily create custom branch policies for environments via the UI.

**Comment by dpuga1** [comment]
How do I do that I’m very new to this?
David Puga

**Reply by dpuga1** [reply]
Can u send me a link?
